src/application/clients/crux-wallet-client.ts

```diff
@@ -0,0 +1,211 @@
+    }
+    return {
+        configurable: true,
+        enumerable: true,
+        get() {
+            if (!patchedFn) {
+                patchedFn = async (...params: any[]) => {
+                    try {
+                        return await fn.call(this, ...params);
+                    } catch (error) {
+                        throw CruxClientError.fromError(error);
+                    }
+                };
+            }
+            return patchedFn;
+        },
+        set(newFn: any) {
+            patchedFn = undefined;
+            fn = newFn;
+        },
+    };
+};
+export interface ICruxWalletClientOptions {
+    privateKey?: string;
+    blockstackInfrastructure?: ICruxBlockstackInfrastructure;
+    cacheStorage?: StorageService;
+    walletClientName: string;
+}
+
+export class CruxWalletClient {
+    public walletClientName: string;
+    private cruxBlockstackInfrastructure: ICruxBlockstackInfrastructure;
+    private _initPromise: Promise<void>;
+    private _cruxUser?: CruxUser;
+    private cruxDomain?: CruxDomain;
+    private _cruxUserRepository: ICruxUserRepository;
+    private _cruxAssetTranslator?: CruxAssetTranslator;
+    private _keyManager?: IKeyManager;
+    private resolvedClientAssetMapping?: IResolvedClientAssetMap;
+    private cacheStorage?: StorageService;
+
+    constructor(options: ICruxWalletClientOptions) {
+        setCacheStorage(options.cacheStorage || new InMemStorage());
+        this.cruxBlockstackInfrastructure = options.blockstackInfrastructure || CruxSpec.blockstack.infrastructure;
+        this.walletClientName = options.walletClientName;
+        this._cruxUserRepository = new BlockstackCruxUserRepository({cacheStorage: this.cacheStorage, blockstackInfrastructure: this.cruxBlockstackInfrastructure});
+        this._initPromise = this._init(options);
+    }
+
+    @throwCruxClientError
+    public getCruxIDState = async (): Promise<ICruxIDState> => {
+        await this._initPromise;
+        if (!this._cruxUser) {
+            if (this._keyManager) {
+                return {
+                    cruxID: null,
+                    status: {
+                        status: "NONE",
+                        statusDetail: "",
+                    },
+                };
+            } else {
+                throw ErrorHelper.getPackageError(null, PackageErrorCode.PrivateKeyRequired);
+            }
+        }
+        return {
+            cruxID: this._cruxUser.cruxID.toString(),
+            status : this._cruxUser.registrationStatus,
+        };
+    }
+
+    @throwCruxClientError
+    public resolveCurrencyAddressForCruxID = async (fullCruxID: string, walletCurrencySymbol: string): Promise<IAddress> => {
+        await this._initPromise;
+        const tag = "resolving_address";
+        const cruxUser = await this._getCruxUserByID(fullCruxID, tag);
+        if (!cruxUser) {
+            throw ErrorHelper.getPackageError(null, PackageErrorCode.UserDoesNotExist);
+        }
```

Comment:
this particular check is repeated in multiple methods. This can be refactored into a private method as `getCruxUser()`
```
if (!this.cruxUser) {
            throw ErrorHelper.getPackageError(null, PackageErrorCode.UserDoesNotExist);
        }
return this.cruxUser
```

Reply:
Only being used once